### PR TITLE
Fix a few profiler related issues associated with frozen objects

### PR DESCRIPTION
**Pull request by cshung**
This change fixes a few profiler-related issues associated with frozen objects.

1. Reset the pinned queue only once
2. Publish frozen objects
~3. Report the frozen segments as survived~
4. Set frozen segment `gen_num`

**Reset the pinned queue only once**

This was an existing bug that is unrelated to frozen objects (for reason I don't understand it only got exposed recently). 

In `walk_relocation`, we walk the generations (from old to young), the associated bricks (in region list order), and the associated plug trees (using an in-order traversal).  During the walk, we dequeue the pinned plugs from the pinned plug queue. 

There is only one pinned queue shared across all generations. Therefore we should not reset the pinned queue for every generation.

**Publish frozen objects**

Some profilers track object instances. If we don't report an object as allocated, they won't be tracking it, and they might be surprised if later on the object is reported as root.

Before the change, frozen objects are not reported as allocated.

After the change, frozen objects allocated through `FrozenObjectHeapManager::TryAllocate` are now reported to the profiler through `PublishObjectAndNotify`, and therefore they will be available through `ICorProfiler::ObjectAllocated` callback.

~**Report the frozen segments as survived**~

~Some profilers might assume that objects previously allocated not falling in that survived range are being collected.~

~Before the change, frozen segments are not reported, and therefore the profiler might assume frozen objects are collected, while they are not, and then got surprised that they got reported as root.~

~After the change, the frozen segments are reported as survived, to do that, the `walk_relocation_sip` method is renamed to `walk_relocation_special_segments` to also handle the special case for frozen segments.~

**Set frozen segment `gen_num`**

As we report the regions to the profiler, we will read the `gen_num` of the heap segment. The current code didn't set that value, so we might be reading random numbers in the profiler. That's not good. As an implementation detail, we thread the frozen segment into gen2, so we might as well set the `gen_num` to 2 there.

@dotnet/gc

### CODE REVIEW

**Comment by cshung** [comment]
https://github.com/dotnet/runtime/issues/75836

**Review by PeterSolMS — APPROVED** [review]
Looks good to me!

**Review by davmason — APPROVED** [review]
LGTM from the profiler side, I am not a GC expert though :)

**Review by EgorBo — APPROVED** [review]

File: src/coreclr/vm/gchelpers.cpp [hunk 1]
```diff
@@ -886,10 +891,12 @@ STRINGREF AllocateString(DWORD cchStringLength, bool preferFrozenHeap, bool* pIs
     if (preferFrozenHeap)
     {
         FrozenObjectHeapManager* foh = SystemDomain::GetFrozenObjectHeapManager();
-        orString = static_cast<StringObject*>(foh->TryAllocateObject(g_pStringClass, totalSize));
+        orString = static_cast<StringObject*>(foh->TryAllocateObject(g_pStringClass, totalSize, /* publish = */false));
         if (orString != nullptr)
         {
             orString->SetStringLength(cchStringLength);
+            // Publish needs to be postponed in this case because we need to specify string length
```
Comment (EgorBo) [inline src/coreclr/vm/gchelpers.cpp:898]: I remember we were discussing this case with @davmason 🙂

**Comment by cshung** [comment]
@Maoni0 and I had a discussion on this and we believe reporting the frozen segments as a plug is inappropriate. We need a different fix for that part.

File: src/coreclr/gc/gcee.cpp [hunk 2]
```diff
@@ -450,6 +450,9 @@ segment_handle GCHeap::RegisterFrozenSegment(segment_info *pseginfo)
     heap_segment_next(seg) = 0;
     heap_segment_used(seg) = heap_segment_allocated(seg);
     heap_segment_plan_allocated(seg) = 0;
+#ifdef USE_REGIONS
+    heap_segment_gen_num(seg) = 2;
```
Comment (Maoni0) [inline src/coreclr/gc/gcee.cpp:454]: as always, let's not use a seemingly random integer.

File: src/coreclr/gc/gc.cpp [hunk 3]
```diff
@@ -33324,22 +33324,20 @@ void gc_heap::walk_relocation (void* profiling_context, record_surv_fn fn)
...
+    reset_pinned_queue_bos();
+    update_oldest_pinned_plug();
+
     for (int i = condemned_gen_number; i >= stop_gen_idx; i--)
     {
         generation* condemned_gen = generation_of (i);
-        heap_segment*  current_heap_segment = heap_segment_rw (generation_start_segment (condemned_gen));
-#ifdef USE_REGIONS
-        current_heap_segment = walk_relocation_sip (current_heap_segment, profiling_context, fn);
+        heap_segment*  current_heap_segment = generation_start_segment (condemned_gen);
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:33333]: I presume you will revert these changes (except keep the `reset_pinned_queue_bos/update_oldest_pinned_plug` part)?

**Review by Maoni0 — APPROVED** [review]
